flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriter.java

```diff
@@ -180,17 +177,33 @@ private void doComplete(boolean precondition, RunnableWithException complete, Ru
 	}
 
 	private <I, H extends AbstractChannelStateHandle<I>> void complete(
+			StreamStateHandle underlying,
 			CompletableFuture<Collection<H>> future,
 			Map<I, List<Long>> offsets,
-			BiFunction<I, List<Long>, H> buildHandle) {
+			TriFunction<I, StreamStateHandle, List<Long>, H> buildHandle) throws IOException {
 		final Collection<H> handles = new ArrayList<>();
 		for (Map.Entry<I, List<Long>> e : offsets.entrySet()) {
-			handles.add(buildHandle.apply(e.getKey(), e.getValue()));
+			handles.add(createHandle(underlying, buildHandle, e));
 		}
 		future.complete(handles);
 		LOG.debug("channel state write completed, checkpointId: {}, handles: {}", checkpointId, handles);
 	}
 
+	private <I, H extends AbstractChannelStateHandle<I>> H createHandle(
+			StreamStateHandle underlying,
+			TriFunction<I, StreamStateHandle, List<Long>, H> buildHandle,
+			Map.Entry<I, List<Long>> e) throws IOException {
+		if (underlying instanceof ByteStreamStateHandle) {
+			ByteStreamStateHandle byteHandle = (ByteStreamStateHandle) underlying;
+			return buildHandle.apply(
+				e.getKey(),
+				new ByteStreamStateHandle(randomUUID().toString(), serializer.extractAndMerge(byteHandle.getData(), e.getValue())),
+				singletonList(serializer.getHeaderLength()));
+		} else {
```

Comment:
It doesn't look like a generic/error prone solution. It looks like something should be extracted to the `StreamStateHandle`. `boolean StreamStateHandle#isDirectlyHoldingData()`? Or we are missing some kind of different abstraction of shared state handles.

Why is this issue new for spilled channel state? What's different for operators state?

Reply:
Added `StreamStateHandle.asBytesIfInMemory` that returns `Optional<byte[]>` (better names are welcomed ;)

> Why is this issue new for spilled channel state? What's different for operators state?

Operators don't share state handles with each other so they don't have this problem.

Reply:
I think it's better now, but I still can not put a finger on a feeling that something might not be entirely correct. 

Why do we have multiple handles per channel, instead of a single handle that has multiple channels?

Reply:
There is only one `InputChannelStateHandle` per channel.
Putting multiple channels could make recovery (rescaling) more difficult.

Reply:
After an offline discussion, we decided to keep it for now as it is. Could you create a JIRA ticket to revisit this issue later (linked to how to handle rescaling) and linked it as `TODO` here?

Reply:
Sure. Created a JIRA ticket: https://issues.apache.org/jira/browse/FLINK-17972